USB Host: integrate class API [2: helpers]

josuah opened the pull request:
Main thread:

- https://github.com/zephyrproject-rtos/zephyr/issues/95661

Downstream:

- https://github.com/zephyrproject-rtos/zephyr/pull/94591
- https://github.com/zephyrproject-rtos/zephyr/pull/99334
- https://github.com/zephyrproject-rtos/zephyr/pull/99173
- https://github.com/zephyrproject-rtos/zephyr/pull/99097
- https://github.com/zephyrproject-rtos/zephyr/pull/99775
- https://github.com/zephyrproject-rtos/zephyr/pull/100703

It contains all the commits of the previous PR, and in addition:

- Wrapper for the class API
- Split the `struct usbh_class_data` and `usbh_class_api`
- Add utilities to register classes as a linked list (NXP)
- Add a filter logic to match a device class to host class instances (NXP)

This also contains functions imported from #94085 by @AidenHu.

So far, this was only tested with this command to make sure it builds and run, but not in hardware yet:
```
west build -t run -b native_sim samples/subsys/usb/shell/ \
-DEXTRA_CONF_FILE=virtual.conf \
-DDTC_OVERLAY_FILE=virtual.overlay \
-DEXTRA_CONF_FILE=device_and_host_prj.conf
```

The PR #94085 got then completely rebased on top of this, available at [`josuah:enable_usb_host_video_class_api2`](https://github.com/josuah/zephyr/tree/enable_usb_host_video_class_api2), which is meant as a drop-in replacement for #94085.

It was also built (but not run) with this command:

```
west build -b rd_rw612_bga -S video-sw-generator samples/drivers/video/capture
```

All tests are done on top of https://github.com/zephyrproject-rtos/zephyr/compare/main...josuah:zephyr:pr_usb_host_class_api2

Commit:
usb: uvc: move the header definition to include/

Move the UVC header with all the definitions from the UVC standard to
share it between USB host and device class implementation.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: add host status to the host context

Add a "struct usbh_status" that contains a bitmask of flags to keep
track of the global state of the host context, like done for the
device_next implementation.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: class: implement more of the class/data struct fields

Add a "struct usbh_class_api" for the host implementation, and move all
the function poitners to it. Add more fields to "struct usbh_class_data".

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: introduce wrappers to access the class function pointers

Add API wrappers around the function pointers in struct usbh_class_api,
while also documenting the USB host class internal API.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: introduce usbh_class with init/remove functions

Add functions to probe/remove all classes as part of a new usbh_class.c
and a matching usbh_class.h. These functions are called from the function
usbh_init_device_intl() in usbh_core.c to initialize every class upon
connection of a device. Every class driver provide filters to match the
interfaces of the device.

Co-authored-by: Aiden Hu <weiwei.hu@nxp.com>
Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

tmon-nordic commented on subsys/usb/host/usbh_device.c:
```diff
@@ -397,7 +397,7 @@ int usbh_device_set_configuration(struct usb_device *const udev, const uint8_t n
 
 	udev->cfg_desc = k_heap_alloc(&usb_device_heap,
-				      cfg_desc.wTotalLength,
+				      cfg_desc.wTotalLength + sizeof(struct usb_desc_header),
```
Can this commit be dropped now that e8ecb6c3af60d30d766271202fb4d41d8bd71cdf is there?

jfischer-no commented on subsys/usb/host/usbh_class.c:
```diff
@@ -57,3 +57,39 @@ int usbh_init_registered_classes(struct usbh_context *uhs_ctx)
 	return 0;
 }
+
+bool usbh_class_is_matching(struct usbh_class_data *cdata,
+			    struct usbh_class_filter *device_info)
```
Please use `const` extensively for the parameters.

```c
bool usbh_class_is_matching(struct usbh_class_data *const cdata,
			    struct usbh_class_filter *const device_info)
```

I did not understand why the name of the second parameter is `device_info`, would not just `filter` be a better name?

josuah replied:
This function needs to know:

- The list of filters configured for that class (`cdata->filters`, passed through `cdata` and not explicitly).
- Information from the device descriptor to match against this list of rules.

I am probably abusing this same struct for these two purpose. Maybe this would make more sense?
```c
bool usbh_class_is_matching(struct usbh_class_filter *filters, size_t num_filters,
                            struct usbh_class_filter *device_info);
```

It got imported into the current UVC PR from NXP which gives an usage example:
https://github.com/zephyrproject-rtos/zephyr/pull/94085/files#diff-46dc301268b060b9975dc2b009b8c30c129b395c64a9c4c12e4e329dc203cd95R108

The device information is first collected from the descriptors:
```c
		if (desc->bDescriptorType == USB_DESC_INTERFACE) {
			struct usb_if_descriptor *if_desc = (void *)desc;

			device_info.code_triple.dclass = if_desc->bInterfaceClass;
			device_info.code_triple.sub = if_desc->bInterfaceSubClass;
			device_info.code_triple.proto = if_desc->bInterfaceProtocol;
			found_interface = true;
		}
```
And then this aggregated info is tested against every class:
```c
		SYS_SLIST_FOR_EACH_CONTAINER(&ctx->class_list, cdata, node) {
			if (usbh_class_is_matching(cdata, &device_info)) {
				LOG_INF("Class driver %s matched for class 0x%02x",
					cdata->name, device_info.code_triple.dclass);
				[...]
			}
		}
```

Is there a better way?

jfischer-no replied:
I would imagine usbh_class_is_matching() should be called from the function driver for the rough check, usbd_class_probe()->usbh_class_is_matching(). How a specific feature of the function is discovered is up to the driver. From your example, all the functions you list should be handled by the same function driver.
I suggest not being UVC specific, or trying to file an API for UVC. Let me take your PR and add a simpler class skeleton to that, as opposed to device_next loopback.c.

josuah replied:
I am connecting this thread with "move the detection logic to the driver and perform the comparison there" [[1]](https://github.com/zephyrproject-rtos/zephyr/pull/94590#discussion_r2325309432), and understands better now thank you.

jfischer-no commented on include/zephyr/usb/usbh.h:
```diff
@@ -111,6 +112,29 @@ struct usbh_class_api {
+	/** Mask of match types for selecting which rules to apply */
+	uint32_t flags;
+	/** The device's class code, subclass code, protocol code. */
+	struct usbh_code_triple code_triple;
```
Maybe just
```c
	/** Device Class Code */
	uint8_t dclass;
	/** Class Subclass Code */
	uint8_t sub;
	/** Class Protocol Code */
	uint8_t proto;
```

Then it looks a bit less painful in usbh_class_is_matching().

josuah commented on include/zephyr/usb/usbh.h:
```diff
@@ -84,28 +86,45 @@ struct usbh_code_triple {
 	/** Request completion event handler */
-	int (*request)(struct usbh_context *const uhs_ctx,
-			struct uhc_transfer *const xfer, int err);
+	int (*request)(struct usbh_class_data *cdata,
```
Naming change `request()` -> `completion_cb()` done.
Let me know if other APIs should also have a `_cb` suffix.

josuah commented on subsys/usb/host/usbh_class.c:
```diff
@@ -0,0 +1,59 @@
+{
+	int registered_count = 0;
+
+	STRUCT_SECTION_FOREACH(usbh_class_data, cdata_existing) {
+		struct usbh_class_data *cdata_registered;
+		bool already_registered = false;
```
Then the whole `subsys/usb/host/usbh_class.c` file can be dropped and this simplifies everything.
I will switch everything from `SYS_SLIST_FOR_EACH_CONTAINER` to just use `STRUCT_SECTION_FOREACH` directly.

Commit:
tests: usb: host: test API using initialization code from USB samples

Add tests making sure the USB Host class APIs introduced build
and run as expected. Use USB device samples common code for
initialization.

Co-authored-by: Johann Fischer <johann.fischer@nordicsemi.no>
Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>
Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

jfischer-no commented on tests/subsys/usb/host/src/main.c:
Looks like it is class API test and should be tests/subsys/usb/host/class_api/src/main.c.

josuah commented on subsys/usb/host/usbh_class_api.h:
```diff
@@ -0,0 +1,186 @@
+ *
+ * @return 0 on success, negative error code on failure.
+ */
+static inline int usbh_class_rwup(struct usbh_class_data *const c_data,
+				  struct usb_device *const udev)
```
I am now back on track with this and should be able to un-block the situation. Sorry to keep you waiting 2 weeks!

josuah commented:
Force-push:
- Add some unit tests
- Apply changes as requested

Still more changes to go...

josuah reviewed (commented):
Here is another cycle.

I believe I addressed most points but if there remain any, I will make sure to be more reactive and allow this to be adjusted quicly.

josuah commented:
I guess the best way for me to discover if the API is good is to try to use it with a class so I will pursue with the rest:

- https://github.com/zephyrproject-rtos/zephyr/pull/94591

josuah commented:
Force-push: from trying to implement a basic class with it:

- Small adjustments that turned out making class implementation easier
- Bugfixes
- Rebase on main.

Waiting CI to go green then un-drafting it.

Commit:
usb: host: add copyright notice to linker script

Add missing copyright notice for the linker script to help with
check_compliance.py.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

MarkWangChinese commented on subsys/usb/host/usbh_class.c:
```diff
@@ -0,0 +1,186 @@
+		} else {
+			__ASSERT(false, "bad implementation of usbh_desc_get_next_function()");
+			return -EINVAL;
+		}
```
This part seems duplicated with the follow calling of `usbh_desc_get_iface_info`.

MarkWangChinese commented on subsys/usb/host/usbh_core.c:
```diff
@@ -71,6 +75,11 @@ static void dev_connected_handler(struct usbh_context *const ctx,
+	}
 }
 
 static void dev_removed_handler(struct usbh_context *const ctx)
```
Will you call `usbh_class_remove_all` in this pr?

josuah replied:
This must have been an omission, thank you for the reminder.

MarkWangChinese reviewed (commented):
Add few comments for some code snippets. I see the comment of https://github.com/zephyrproject-rtos/zephyr/pull/94085#discussion_r2444793774 too. Just one idea for discussion: Is it OK to use the device tree to define the rule to limit the matched usb class function.
For example:
```
	usb_hid_mouse {
		compatible = "zephyr,uhid-host";
		protocol = "mouse"; // the usb host hid class driver make sure only the mouse interface match with this instance.
	};

	usb_hid_keyboard {
		compatible = "zephyr,uhid-host";
		protocol = "keyboard"; // the usb host hid class driver make sure only the keyboard interface match with this instance.
	};

        lvgl_pointer {
                compatible = "zephyr,lvgl-pointer-input";
                input = <&usb_hid_mouse>;
        };

        lvgl_keypad_input {
                compatible = "zephyr,lvgl-keypad-input";
                input = <&usb_hid_keyboard>;
        };
```
For the mass, I think it is hard to define the rule. I think It needs the application to use the file system (lfs) to check which partition is used.
```
        fstab {
                compatible = "zephyr,fstab";

                lfs0: lfs0 {
                        compatible = "zephyr,fstab,littlefs";
                        partition = <&usb_msc0_partition>;
                        mount-point = "/lfs0";
                };

                lfs1: lfs1 {
                        compatible = "zephyr,fstab,littlefs";
                        partition = <&usb_msc1_partition>;
                        mount-point = "/lfs1";
                };
        };
```

GeorgeCGV commented on subsys/usb/host/usbh_core.c:
```diff
@@ -71,6 +75,11 @@ static void dev_connected_handler(struct usbh_context *const ctx,
+
+	ret = usbh_class_probe_all(ctx->root);
+	if (ret != 0) {
+		LOG_ERR("Failed to probe all classes for this new USB device");
```
```suggestion
		LOG_ERR("Failed to probe classes for new USB device, %d", ret);
```

GeorgeCGV commented on subsys/usb/host/usbh_desc.c:
```diff
@@ -0,0 +1,179 @@
+				   uint8_t *const iface_num)
+{
+	if (desc->bDescriptorType == USB_DESC_INTERFACE_ASSOC) {
+		struct usb_association_descriptor *iad_desc = (void *)desc;
```
Why not?
```suggestion
		const struct usb_association_descriptor *iad_desc = (const struct usb_association_descriptor *)desc;
```

GeorgeCGV commented on subsys/usb/host/usbh_desc.h:
```diff
@@ -0,0 +1,132 @@
+ * @brief Get the next descriptor in an array of descriptors.
+ *
+ * This
+ * to search either an interface or  interface association descriptor:
```
Abrupted "This"?

GeorgeCGV commented on subsys/usb/host/usbh_class.c:
```diff
@@ -0,0 +1,186 @@
+		if (c_node->c_data->udev == udev) {
+			ret = usbh_class_removed(c_node->c_data);
+			if (ret != 0) {
+				LOG_ERR("Failed to handle device removal for each classes");
```
```suggestion
				LOG_ERR("Failed to handle device removal for each class, %d", ret);
```

josuah replied:
done

josuah commented on subsys/usb/host/usbh_desc.c:
```diff
@@ -0,0 +1,179 @@
+
+	if (desc->bDescriptorType == USB_DESC_INTERFACE_ASSOC) {
+		iface = iad_desc->bFirstInterface;
+		skip_num = iad_desc->bInterfaceCount;
```
This is a good point, this will need handling of this case to give a consistent API.
Thank you.

josuah commented on subsys/usb/host/usbh_class.h:
```diff
@@ -0,0 +1,67 @@
+ * @brief Match an USB host class (a driver) against a device descriptor.
+ *
+ * @param[in] filters Array of filter rules to match
+ * @param[in] num_filters Number of rules in the array.
```
done

josuah commented on subsys/usb/host/usbh_class.c:
```diff
@@ -0,0 +1,186 @@
+	bool matching = false;
+	int ret;
+
+	STRUCT_SECTION_FOREACH(usbh_class_node, c_node) {
```
done

MarkWangChinese replied:
OK, it seems that the codes don't consider the composite device? For example: One attached usb device has both video and hid function. If the first video function is matched successfully, then usbh_class_probe_all return, then hid function is not used.
My suggestion: put the attached device's functions (IAD, interfaces) as the first loop, then check all the idle class instance nodes to match with the functions as the second loop.

josuah commented on subsys/usb/host/usbh_class.h:
```diff
@@ -0,0 +1,67 @@
+ */
+
+#ifndef ZEPHYR_INCLUDE_USBD_CLASS_H
+#define ZEPHYR_INCLUDE_USBD_CLASS_H
```
done

josuah commented on subsys/usb/host/usbh_desc.h:
```diff
@@ -0,0 +1,132 @@
+const void *usbh_desc_get_next(const void *const desc_beg, const void *const desc_end);
+
+/**
+ * @brief Search the first descriptor matching the selected type(s).
```
done

josuah commented on subsys/usb/host/usbh_class.h:
```diff
@@ -0,0 +1,67 @@
+int usbh_class_init_all(struct usbh_context *const uhs_ctx);
+
+/**
+ * @brief Probe all classes to against a newly connected USB device.
```
done

josuah commented on subsys/usb/host/usbh_class_api.h:
```diff
@@ -0,0 +1,158 @@
+/**
+ * @brief Bus resumed handler
+ *
+ * Called when the host resumes the activity on a bus.
```
done

josuah commented on subsys/usb/host/usbh_core.c:
```diff
@@ -194,12 +203,10 @@ int usbh_init_device_intl(struct usbh_context *const uhs_ctx)
-		break;
+	ret = usbh_class_init_all(uhs_ctx);
+	if (ret != 0) {
+		LOG_ERR("Failed to initialize all classes");
```
done

josuah commented on subsys/usb/host/usbh_class.c:
```diff
@@ -0,0 +1,186 @@
+	STRUCT_SECTION_FOREACH(usbh_class_node, c_node) {
+		ret = usbh_class_init(c_node->c_data, uhs_ctx);
+		if (ret != 0) {
+			LOG_ERR("Failed to initialize all registered class instances");
```
done

josuah commented on subsys/usb/host/usbh_class.c:
```diff
@@ -0,0 +1,186 @@
+	}
+
+	if (!matching) {
+		LOG_WRN("Could not find any class for this device");
```
done

josuah commented:
> Littlefs is not particurarly a wise choice for USB Mass Storage. MSC operates on block device, so if there is wear leveling involved should be Flash Translation Layer on the device side and not on the filesystem level.

I agree, I do not know why I gave this example, I probably meant to point at FAT FS.

GeorgeCGV commented on subsys/usb/host/usbh_desc.h:
```diff
@@ -0,0 +1,106 @@
+				   uint8_t *const iface_num);
+
+/**
+ * Checks that the pointed descriptor is not truncated.
```
Missing `@brief`?

GeorgeCGV replied:
Ok, then shouldn't internal parts be wrapped with `/** @cond INTERNAL_HIDDEN */` and `/** INTERNAL_HIDDEN @endcond */`?

josuah commented:
Converted to draft until I fix CI...

tmon-nordic commented on subsys/usb/host/usbh_desc.c:
```diff
@@ -0,0 +1,215 @@
+	return -EINVAL;
+}
+
+const void *usbh_desc_get_next_function(const void *const desc_beg, const void *const desc_end)
```
>This way, it is possible to jump from function to function, and i.e. try to assign them to a class.
>
> Does it make sense?

Yes, this is precisely what IAD is for.

josuah commented:
Force-push:

- Fix CI issues
- Documentation fixes
- Simplified descriptor utilities
- More descriptor parsing utilities
- Always check descriptor validity (safe use)
- Change the way probe order works (`for each function, { for each class  { .. } }`)
- Check that a function is not already used by a class before trying to bind it to a driver.

josuah commented:
@AidenHu good idea to try to make some "aggregation" pull requests to test things together before they are merged.

I prefer to add it to a new API4 so that API2 can be kept small and quicker to review.

Then you can use API4 which will have everything.

Does that make sense?

AidenHu commented:
> > * [Support multiple devices to connect with usb host if usb hub is used #99591 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/99591#issuecomment-3557914320)
> 
> @AidenHu good idea to try to make some "aggregation" pull requests to test things together before they are merged.
> 
> I prefer to add it to a new API4 so that API2 can be kept small and quicker to review.
> 
> Then you can use API4 which will have everything.

@josuah  Thank you for the comment. Maybe my description about testing is not clear and let me clarify more here. The testing for API2 can be done by the [host_hub ](https://github.com/zephyrproject-rtos/zephyr/pull/99735). The [host_hub ](https://github.com/zephyrproject-rtos/zephyr/pull/99735) is aimed for implementation for hub class. The [host_hub ](https://github.com/zephyrproject-rtos/zephyr/pull/99735) will always be kept newest head of your current API2 and will always be kept testing status by me. Therefore, during the testing if there is something need to be updated in API2, I will create a commit like [#99591](https://github.com/zephyrproject-rtos/zephyr/pull/99591) or raise comment in API2.  Certainly, I need to make sure the update should be   necessary for API2 so that avoid increasing review effort in API2.

Now for [ #99591](https://github.com/zephyrproject-rtos/zephyr/pull/9959), I think we should not move it to API4, it should be considered now by API2, because it is critical and very necessary to avoid the big API change in the future.

josuah commented on subsys/usb/host/usbh_class.c:
```diff
@@ -0,0 +1,182 @@
+		if (c_node->state == USBH_CLASS_STATE_BOUND &&
+		    c_data->udev == udev && c_data->iface == iface) {
+			LOG_DBG("Interface %u bound to '%s', skipping", iface, c_data->name);
+			return;
```
Here and at another place, so I reworked the locking to be more consistent. Thanks.

josuah commented:
Force-push:
- Address review comments

tmon-nordic commented on subsys/usb/host/usbh_class.h:
```diff
@@ -0,0 +1,78 @@
+#define USBH_CLASS_MATCH_SUB BIT(4)
+
+/** Match a protocol code */
+#define USBH_CLASS_MATCH_PROTO BIT(5)
```
The individual criteria seem a bit odd. Matching on code triple is fine, but on individual fields seems questionable.

Matching on VID alone is very questionable, while matching on PID alone sounds outright wrong.

josuah replied:
That makes sense. Reduced to a combined `_VID_PID` and `_CODE_TRIPLE` only/

tmon-nordic commented on subsys/usb/host/usbh_desc.h:
```diff
@@ -0,0 +1,149 @@
+ *
+ * @return A pointer to the beginning of the descriptor
+ */
+const void *usbh_desc_get_cfg_beg(const struct usb_device *const udev);
```
Why not `usbh_desc_get_cfg`?

josuah replied:
Just for symmetry with `usbh_desc_get_cfg_end` and to remind users they need to call both in combination for safety purpose. However it is enforced by all APIs...

Removed the `_beg` suffix.

tmon-nordic commented on subsys/usb/host/usbh_desc.h:
```diff
@@ -0,0 +1,149 @@
+ *
+ * @param[in] udev The device holding the configuration descriptor
+ *
+ * @return A pointer to the beginning of the descriptor
```
Copy&paste error. THis function is supposed to return the end.

tmon-nordic commented on subsys/usb/host/usbh_desc.h:
```diff
@@ -0,0 +1,149 @@
+ * If @p desc_beg is another type of descriptors, it will seek to a matching descriptor type and
+ * return it, without skipping to the next one after it.
+ *
+ * @param[in] desc_beg Pointer to the beginning of the descriptor array; to search. May be NULL.
```
A bit strange name. I would go with just `desc`.

josuah commented on subsys/usb/host/usbh_class.c:
```diff
@@ -0,0 +1,197 @@
+			    struct usbh_class_filter *const info)
+{
+	/* Make empty filter set match everything (use class_api->probe() only) */
+	if (filters[0].flags == 0) {
```
That is a better way to make a distinction... It also makes it possible to write classes that match nothing, which is safer and easier to debug if i.e. every rule is commented out by `#ifdef CONFIG_...`.

tmon-nordic commented on subsys/usb/host/usbh_class_api.h:
```diff
@@ -0,0 +1,147 @@
+		return api->suspended(c_data);
+	}
+
+	return -ENOTSUP;
```
I didn't mean to remove the suspend callback, but rather to allow it to be optional, i.e. to not return `-ENOTSUP`, but just 0 if there is no callback.

Commit:
usb: host: add debug logs for the configuration descriptor

The USB host will have to select the appropriate size for the configurtion
descriptor, and the defaults will not be enough for some classes that
encode a lot of information in the configuration descriptor.
Add logging to debug this.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

jfischer-no commented on subsys/usb/host/usbh_desc.h:
```diff
@@ -0,0 +1,149 @@
+ * @param[out] iface_code Device information filled by this function
+ * @param[out] iface_num Pointer filled with the interface number, or NULL
+ *
+ * @return 0 on success or negative error code on failure.
```
What are positive values? Please replace all the `if (ret < 0)` with `if (ret != 0)`

jfischer-no commented on subsys/usb/host/usbh_class.c:
```diff
@@ -0,0 +1,188 @@
+
+			/* The class instance is now free to bind to a new device */
+			c_data->udev = NULL;
+			c_node->state = USBH_CLASS_STATE_IDLE;
```
I think it is not the right way to manage resources and likely will be difficult to rework later.

jfischer-no commented on subsys/usb/host/usbh_class.c:
```diff
@@ -0,0 +1,188 @@
+	k_mutex_lock(&udev->mutex, K_FOREVER);
+
+	desc = usbh_desc_get_cfg(udev);
+	desc_end = usbh_desc_get_cfg_end(udev);
```
Drop desc_end and usbh_desc_get_cfg_end()

josuah replied:
done

jfischer-no commented on subsys/usb/host/usbh_class.c:
```diff
@@ -0,0 +1,188 @@
+	uint8_t iface;
+	int ret;
+
+	k_mutex_lock(&udev->mutex, K_FOREVER);
```
This function must be called from a single place in usbh_core, in a host thread context. Is k_mutex_lock() really necessary here?

josuah commented:
Thank you for the review! I will try to address everything this afternoon.

josuah commented:
- remove use of `desc_end`
- convert from `ret < 0` to `ret != 0`

josuah commented on subsys/usb/host/usbh_desc.c:
```diff
@@ -0,0 +1,214 @@
+const void *usbh_desc_get_cfg(const struct usb_device *const udev)
+{
+	return udev->cfg_desc;
+}
```
Maybe even removing the function altogether which just accesses a field now that  there is no `usbh_desc_get_cfg_end()`?

josuah replied:
done

josuah commented on subsys/usb/host/usbh_desc.c:
```diff
@@ -0,0 +1,214 @@
+	/* Avoid too short descriptors */
+	if ((const uint8_t *)desc + expected_size > (const uint8_t *)desc_end) {
+		return false;
+	}
```
done, thank you

Commit:
usb: host: improve safety of configuration descriptor parsing

Make sure that there is at least enough room for the header before
dereferencing the fields such as bLength.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
usb: host: add a nil entry at the end of configuration descriptor

Allocate a larger size to keep room for a nil descriptor at the end of
the configuration descriptor. This is used to walk through the list of
descriptors until this empty descriptor happens.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

josuah commented:
- Coding style adjustments

josuah commented on subsys/usb/host/usbh_desc.c:
```diff
@@ -0,0 +1,214 @@
+{
+	if (desc_ptr == NULL || desc_end == NULL) {
+		return NULL;
+	}
```
done, check removed

jfischer-no committed:
usb: host: add helpers to parse USB descriptors

Add helpers to parse and validate USB descriptors.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>
Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

jfischer-no committed:
usb: host: add functions to initialize and probe class drivers

Add basic functions to initialize and probe class drivers.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>
Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

AidenHu commented on subsys/usb/host/usbh_desc.c:
```diff
@@ -0,0 +1,188 @@
+	}
+
+	return head;
+}
```
Hi @josuah 
I found `usbh_desc_get_next_function` does not handle device's interface properly. Take an example, device cdc ecm has a communication interface and data interface, here two interfaces will be regarded as two independent function, then probe function will be called twice but device cdc ecm is not a composite device. 
Or do we need to adjust filter usage to fix this issue?

jfischer-no commented on subsys/usb/host/usbh_device.c:
```diff
@@ -424,7 +424,8 @@ int usbh_device_set_configuration(struct usb_device *const udev, const uint8_t n
 	if (err) {
-		LOG_ERR("Failed to read configuration descriptor");
+		LOG_ERR("Failed to read configuration descriptor of %u bytes: %d",
+			cfg_desc.wTotalLength, err);
```
>Every USB host class and platform will have to select the appropriate
size for the configurtion descriptor, and the defaults will not be enough
for some classes that encode a lot of information in the configuration
descriptor.

No objections to log it, but class drivers have nothing to select here. It is unusual for this request to fail and a device may return a configuration descriptor that has no meaning for any class driver. If there is not enough memory then it will fail and log earlier.

josuah replied:
done, I got confused in the git message, adjusted

jfischer-no commented on tests/subsys/usb/host/class_api/src/main.c:
```diff
@@ -0,0 +1,318 @@
+};
+
+struct usb_device test_udev = {
+	.cfg_desc = (void *)test_hub_descriptor,
```
Not even close to fake a device, .ifaces are missed.

Commit:
usb: host: store the pointers to the interface association

In addition to storing the interface pointers, also store the pointers
for the interface association. Also add comments to the other fields
of the structs to adapt to a new CI check.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

josuah commented:
Force-push, from out-of-github review
- Add a mechanism for matching against the whole device for single-function devices without IAD [WIP]
- Collect IADs in an array and use it for matching [WIP]
- "info" -> "filter"
- tests: do not use a class instance [WIP]

jfischer-no commented on subsys/usb/host/usbh_desc.c:
```diff
@@ -0,0 +1,197 @@
+			*iface = ia_desc->bFirstInterface;
+		}
+		return 0;
+	}
```
Not sure how that could be used. It is only the function driver what would check the device descriptor, number of interfaces and then look for IAD on its own.

josuah replied:
This is being used for applying the filter rules in `subsys/usb/host/usbh_class.c`:

```c
        while (true) {
                desc = usbh_desc_get_next_function(desc);
                if (desc == NULL) {
                        break;
                }

                ret = usbh_desc_fill_filter(desc, &filter_data, &iface);
                if (ret != 0) {
                        LOG_ERR("Failed to collect class codes for matching interface %u",
                                iface);
                        continue;
                }

                usbh_class_probe_function(udev, &filter_data, iface);
        }
```

Where `usbh_class_probe_function()` uses the extracted `filter_data`:

```c
                if (!usbh_class_is_matching(c_node->filters, filter_data)) {
                        LOG_DBG("Class %s not matching interface %u",
                                c_data->name, iface);
                        continue;
                }

                ret = usbh_class_probe(c_data, udev, iface);
                if (ret == -ENOTSUP) {
                        LOG_DBG("Class %s not supporting this function, skipping",
                                c_data->name);
                        continue;
                }
```

Should it be `class->api->probe()` implementation which calls `usbh_class_is_matching()`?

In which case, `struct class_data.filter` can be removed and use of the filter API can be an implementation detail.

jfischer-no commented on subsys/usb/host/usbh_device.c:
```diff
@@ -272,7 +272,7 @@ int usbh_device_interface_set(struct usb_device *const udev,
 }
 
-static int parse_configuration_descriptor(struct usb_device *const udev)
+int usbh_device_parse_cfg_desc(struct usb_device *const udev)
```
This function should not be public. Nothing should call it from outside.

jfischer-no commented on include/zephyr/usb/usbh.h:
```diff
@@ -1,7 +1,6 @@
- * Copyright 2025 NXP
- *
+ * SPDX-FileCopyrightText: Copyright Nordic Semiconductor ASA
+ * SPDX-FileCopyrightText: Copyright 2025 NXP
```
I suggest doing it only when the copyright line is touched, otherwise, it is a bit annoying.

josuah replied:
done

josuah commented on tests/subsys/usb/host/class_api/src/main.c:
```diff
@@ -0,0 +1,318 @@
+	LOG_INF("Host controller enabled");
+
+	/* Allow the host time to reset the host. */
+	k_msleep(200);
```
done

josuah replied:
I removed it but due to this https://github.com/zephyrproject-rtos/zephyr/pull/94590#discussion_r2695529583 I am adding it back.

josuah commented on tests/subsys/usb/host/desc/src/main.c:
```diff
@@ -0,0 +1,73 @@
+ZTEST(host_desc, test_desc_get_by_type)
+{
+	const uint8_t *const desc_beg = test_hub_descriptor;
+	const struct usb_desc_header *desc;
```
done

josuah commented on subsys/usb/host/usbh_device.c:
```diff
@@ -289,6 +290,14 @@ static int parse_configuration_descriptor(struct usb_device *const udev)
+			} else {
+				udev->iads[tmp_nia] = dhp;
+				tmp_nia++;
+			}
```
done, imported into this PR (about to force-push it)

josuah commented on tests/subsys/usb/host/class_api/src/main.c:
```diff
@@ -0,0 +1,296 @@
+USBH_CONTROLLER_DEFINE(uhs_ctx, NULL);
+
+#define TEST_VID 0x2FE3
+#define TEST_PID 0x0002
```
done, as part of the test USB descriptors

josuah commented on tests/subsys/usb/host/class_api/src/main.c:
```diff
@@ -0,0 +1,296 @@
+ZTEST(host_class, test_class_matchiing)
+{
+	struct usb_device *const udev = &test_udev;
+	const struct usb_desc_header *desc = udev->cfg_desc;
```
done

josuah commented on tests/subsys/usb/host/common/include/test_descriptor.h:
```diff
@@ -0,0 +1,86 @@
+	0,		/* iManufacturer */ \
+	0,		/* iProduct */ \
+	0,		/* iSerial */ \
+	1,		/* bNumConfigurations */
```
I will revert my changes in favor of this latest https://github.com/zephyrproject-rtos/zephyr/pull/94590#discussion_r2698314937
AFAIU, this deprecates this suggestion so marking as resolved.

josuah replied:
done

josuah commented on tests/subsys/usb/host/class_api/src/main.c:
```diff
@@ -0,0 +1,296 @@
+	struct usb_device *const udev = &test_udev;
+	int ret;
+
+	ret = usbh_device_parse_cfg_desc(udev);
```
I will revert my changes in favor of this latest https://github.com/zephyrproject-rtos/zephyr/pull/94590#discussion_r2698314937
AFAIU, this deprecates this suggestion so marking as resolved.

josuah replied:
done

josuah commented:
@jfischer-no I applied all review comments except this one: https://github.com/zephyrproject-rtos/zephyr/pull/94590#discussion_r2695470539 as I am not sure what to do.

I went with a different approach for the tests, using UVB, and a empty test class calling `zassert()`. This way, this tests the class API.

josuah commented on include/zephyr/usb/usbh.h:
```diff
@@ -115,8 +116,10 @@ struct usbh_class_data {
-	struct usbh_code_triple code;
+	/** Pointer to USB device this class is used for */
+	struct usb_device *udev;
+	/** Interface number for which this class matched */
```
done

josuah commented on subsys/usb/host/usbh_desc.c:
```diff
@@ -0,0 +1,215 @@
+	return NULL;
+}
+
+const void *usbh_desc_get_endpoint(const struct usb_device *const udev, const uint8_t ep_addr)
```
done, here as well as in `subsys/usb/host/usbh_desc.h`

josuah commented on tests/subsys/usb/host/src/main.c:
```diff
@@ -0,0 +1,395 @@
+		zassert_ok(ret, "Failed to add configuration (%d)", ret);
+	}
+
+	ret = usbd_register_all_classes(&test_usbd, USBD_SPEED_FS, 1, NULL);
```
done, thank you for the example usage

josuah replied:
this is done as well

MarkWangChinese commented on subsys/usb/host/usbh_class.c:
```diff
@@ -0,0 +1,200 @@
+	STRUCT_SECTION_FOREACH(usbh_class_node, c_node) {
+		struct usbh_class_data *const c_data = c_node->c_data;
+
+		if (c_data->udev == udev) {
```
```suggestion
		if (c_data->udev != udev) {
```

MarkWangChinese reviewed (commented):
We do some basic tests with video, hub and ecm PRs, and find the follow two issues. Please check.

josuah commented on subsys/usb/host/usbh_desc.c:
```diff
@@ -0,0 +1,215 @@
+	/* Skip all interfaces the Association descriptor contains */
+	if (usbh_desc_is_valid_association(head)) {
+		skip_num = ass_d->bInterfaceCount;
+	}
```
The tests on tests/subsys/usb/host use `device_next`'s loopback class to provide the tests, and it has an IAD so this bug was not caught. I think it was planned to be more descriptors for stress testing the parsing as a future PR.

I currently pushed the changes.

jfischer-no commented on subsys/usb/host/usbh_desc.c:
```diff
@@ -0,0 +1,218 @@
+	if (usbh_desc_is_valid_association(head)) {
+		skip_num = ass_d->bInterfaceCount;
+	/* Skip the interface if the head is interface */
+	} else if (usbh_desc_is_valid_interface(head)) {
```
It is the same head what is either IAD or Interface descriptor. Please use `} else if {` where it is really necessary. I also would like to avoid "No-action-required" :tm: people be triggered just because else is missed.

```c
	/* Skip all interfaces the Association descriptor contains */
	if (usbh_desc_is_valid_association(head)) {
		skip_num = ass_d->bInterfaceCount;
	}

	/* Skip the interface if the head is interface */
	if (usbh_desc_is_valid_interface(head)) {
```
